src/main/java/com/google/sps/data/Journey.java

```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
+  public ArrayList<Point> findOptimalOrderForWaypoints() throws ApiException, DataNotFoundException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = findOptimalTimePaths();
+    ArrayList<Point> orderedWaypoints = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    // the last state of the matrix has all waypoints set to 1
+    int currentState =  (1 << numberOfWaypoints) - 1;
+    
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
+    // currentWaypoint is the one for which bestTime[currentState][currentWaypoint] is minimal
+    int currentWaypoint = 0;
+    for (int waypointIndex = 1; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(currentState).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(currentState).get(waypointIndex).first;
+        currentWaypoint = waypointIndex;
+      }
+    }
+    while (currentState != 0) {
+      orderedWaypoints.add(waypoints.get(currentWaypoint));
+      int prevState = currentState ^ (1 << currentWaypoint);
+      int prevWaypoint = bestTime.get(currentState).get(currentWaypoint).second;
+      currentWaypoint = prevWaypoint;
+      currentState = prevState;
+    }
+    // add the start of the journey at the end of orderedWaypoints
+    orderedWaypoints.add(start);
+    return orderedWaypoints;
+  }
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config <= (1 << numberOfWaypoints); ++config) {
+      // add the config-th line to the matrix with numberOfWaypoints columns of Integer.MAX_VALUE
+      ArrayList<Pair> currentRow = new ArrayList<Pair>();
+
+      for (int waypointIndex = 0; waypointIndex <= numberOfWaypoints; ++ waypointIndex) {
+        currentRow.add(new Pair(Integer.MAX_VALUE, -1));
+      }
+      bestTime.add(currentRow);
+    }
+
+    int waypointIndex = 0;
+    for (Point waypoint : waypoints) {
+      Pair currentElement = bestTime.get(1 << waypointIndex).get(waypointIndex);
+      currentElement.first = pathFinder.getTimeInSecondsBetweenPoints(start, waypoint);
+      waypointIndex += 1;
+    }
+
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint] is the min
+   * time needed to visit all points with bits set to 1 in config such that the path ends in point
+   * lastWaypoint.
```

Comment:
Can you add some details here on how this computation is performed?

Reply:
Done.